pkg/sbom/io/decode.go

```diff
@@ -348,18 +341,37 @@ func (m *Decoder) addOSPkgs(sbom *types.SBOM) {
 // addLangPkgs traverses relationships and adds language-specific packages
 func (m *Decoder) addLangPkgs(sbom *types.SBOM) {
 	for id, app := range m.apps {
-		for _, rel := range m.bom.Relationships()[id] {
-			pkg, ok := m.pkgs[rel.Dependency]
-			if !ok {
-				continue
-			}
-			app.Packages = append(app.Packages, *pkg)
-			delete(m.pkgs, rel.Dependency) // Delete the added package
-		}
+		app.Packages = append(app.Packages, m.traverseDependencies(id, set.New[uuid.UUID]())...)
 		sbom.Applications = append(sbom.Applications, *app)
 	}
 }
 
+// traverseDependencies recursively retrieves all packages that the specified component depends on.
+// It starts from the given component ID and traverses the dependency tree, collecting all
+// dependent packages. The collected packages are removed from m.pkgs to prevent duplicate
+// processing. This ensures that all dependencies, including transitive ones, are properly
+// captured and associated with their parent component.
+func (m *Decoder) traverseDependencies(id uuid.UUID, seen set.Set[uuid.UUID]) ftypes.Packages {
+	var pkgs ftypes.Packages
+	for _, rel := range m.bom.Relationships()[id] {
+		if seen.Contains(rel.Dependency) {
+			continue // Skip if already seen to avoid infinite loop and duplicate processing
+		}
```

Comment:
I checked case with infinity loop.
We delete package from `m.pkgs`:
https://github.com/aquasecurity/trivy/blob/b9e85e731173ce981ce93573d31e82f3fb4d0525/pkg/sbom/io/decode.go#L367

Therefore, the following logic will avoid infinite loops:
https://github.com/aquasecurity/trivy/blob/b9e85e731173ce981ce93573d31e82f3fb4d0525/pkg/sbom/io/decode.go#L360-L363

Or am i missing something?

Reply:
Oh, you're right. I was smarter in the past 😄 Reverted https://github.com/aquasecurity/trivy/pull/8144/commits/4db29eb054a7d8291bf9914b0827bd8e2572c265